src/components/Requirements/SubRequirement.vue

```diff
@@ -194,8 +184,39 @@ export default Vue.extend({
         Object.keys(this.subReq.requirement.fulfillmentOptions)[0]
       );
     },
-    subReqCoursesArray(): SubReqCourseSlot[] {
-      return this.generateSubReqCoursesArray();
+    subReqCoursesSlots(): SubReqCourseSlot[] {
+      const subReqSpec = getMatchedRequirementFulfillmentSpecification(this.subReq.requirement, {
+        [this.subReq.requirement.id]: this.toggleableRequirementChoice,
+      });
+      if (subReqSpec === null) return [];
+      const subReqEligibleCourses = subReqSpec.eligibleCourses;
+
+      const allTakenCourseIds = new Set(this.subReq.courses.flat().map(course => course.courseId));
+      const slots: SubReqCourseSlot[] = [];
+
+      const coursesTaken = this.subReq.courses;
+
+      coursesTaken.forEach((subReqCourseSlot, i) => {
+        if (subReqCourseSlot.length > 0) {
+          subReqCourseSlot.forEach(subReqCourse => {
+            slots.push({ isCompleted: true, courses: [subReqCourse] });
+          });
+          // Create new IncompletedSubReqCourse slot if all credits or courses not met
+          // but only one CompletedSubReqCourse slot exists
+          if (coursesTaken.length === 1 && !this.isCompleted) {
+            slots.push({
+              isCompleted: false,
+              courses: generateSubReqIncompleteCourses(allTakenCourseIds, subReqEligibleCourses[i]),
+            });
+          }
+        } else {
+          slots.push({
+            isCompleted: false,
+            courses: generateSubReqIncompleteCourses(allTakenCourseIds, subReqEligibleCourses[i]),
+          });
+        }
+      });
+      return slots;
```

Comment:
we no longer need to batch fetch data, so now the resulting flow is much cleaner!